pilot/pkg/model/push_context.go

```diff
@@ -985,6 +1048,61 @@ func (ps *PushContext) initDestinationRules(env *Environment) error {
 	return nil
 }
 
+// AuthenticationPolicyForWorkload returns the matching auth policy for a given service
+// This replaces store.AuthenticationPolicyForWorkload
+func (ps *PushContext) AuthenticationPolicyForWorkload(service *Service, l labels.Instance, port *Port) *authn.Policy {
+	var workloadPolicy *authn.Policy
+
+	// Match by Service hostname
+	workloadPolicy = authenticationPolicyForWorkload(ps.AuthNPolicies[service.Hostname], service, l, port)
+
+	// Match by namespace
+	if workloadPolicy == nil {
+		workloadPolicy = authenticationPolicyForWorkload(ps.AuthNPolicies[host.Name(service.Attributes.Namespace)], service, l, port)
+	}
+
+	// Use default global authentication policy
+	if workloadPolicy == nil {
+		workloadPolicy = ps.defaultAuthNMeshPolicy
+	}
+
+	return workloadPolicy
+}
+
+func authenticationPolicyForWorkload(specsByPort map[int][]*Config, service *Service, l labels.Instance, port *Port) *authn.Policy {
+	var matchedPolicy *authn.Policy
+	if specsByPort == nil {
+		return matchedPolicy
+	}
+	if matchedPolicy = matchPolicyByLabels(specsByPort[port.Port], service, l); matchedPolicy == nil {
+		matchedPolicy = matchPolicyByLabels(specsByPort[noPortProvided], service, l)
+	}
+	return matchedPolicy
+}
+
+func matchPolicyByLabels(specs []*Config, service *Service, l labels.Instance) *authn.Policy {
+	var matchedPolicy *authn.Policy
+	for _, spec := range specs {
+		if spec.Namespace != service.Attributes.Namespace {
```

Comment:
This function was called with the specs that already match service fqdn, so it's not necessary to check namespace again.

You can also have the vector that hold Policy, instead of Config  object

Reply:
@diemtvu FQDN does not imply a known namespace in all cases though.  ServiceEntries can specify DNS names which have no ties to a specific namespace.  The existing configstore implementation has logic for retrieving auth policy configs for only a specific namespace:
https://github.com/istio/istio/blob/master/pilot/pkg/model/config.go#L550.
```
func (store *istioConfigStore) AuthenticationPolicyForWorkload(service *Service, l labels.Instance, port *Port) *Config {
	if len(service.Attributes.Namespace) == 0 {
		return nil
	}
	namespace := service.Attributes.Namespace
	specs, err := store.List(schemas.AuthenticationPolicy.Type, namespace)
```

My implementation uses the namespace from the `*Service` because that service comes from push_context.Services() which has already filtered the services to those visible to that Proxy:
```
// Services returns the list of services that are visible to a Proxy in a given config namespace
func (ps *PushContext) Services(proxy *Proxy) []*Service {
``` 

Even if we know at the start of `initAuthNPolicies` which namespaces are visible to the Proxy and filter them there - we could still potentially expose a different service to an authN policy in a different namespace

Reply:
authn policies do not allow FQDNs for targets, but so the cached map index is generated during the `initAuthN` function using `ResolveShortnameToFQDN`.  So the cached map should be indexed by a hostname which always contain a FQDN derived from it's namespace.  Marking resolved